File: Cpp/src/estimate_abc.cpp
#include <multiRL/estimate_abc.hpp>

#include <multiRL/process_MDP_free.hpp>
#include <multiRL/task_builder.hpp>

#include <abcpp/abc.hpp>
#include <abcpp/abcpp_impl.hpp>
#include <abcpp/options.hpp>
#include <abcpp/summary.hpp>

#include <algorithm>
#include <cmath>
#include <limits>
#include <numeric>
#include <random>
#include <stdexcept>

#ifdef _OPENMP
#include <omp.h>
#endif

namespace multiRL {
namespace {

double finite_or(double value, double fallback) {
    return std::isfinite(value) ? value : fallback;
}

std::string task_subid(const RunTask& task) {
    if (!task.input.idinfo.empty() && !task.input.idinfo[0].empty()) {
        return task.input.idinfo[0][0];
    }
    return "1";
}

std::vector<double> observed_summary(const RunTask& task) {
    const std::size_t n_cues = task.behrule.cue.size();
    std::vector<double> counts(n_cues, 0.0);
    double reward_sum = 0.0;
    double valid_reward = 0.0;

    for (std::size_t row = 0; row < task.input.n_rows; ++row) {
        const std::string& action = task.input.action[row];
        const auto cue_it = task.behrule.cue_index.find(action);
        if (cue_it != task.behrule.cue_index.end()) {
            counts[cue_it->second] += 1.0;
        }

        for (std::size_t option = 0; option < task.input.state[row].size();
             ++option) {
            bool found = false;
            for (const std::string& value : task.input.state[row][option]) {
                if (value == action) {
                    found = true;
                }
            }
            if (found && option < task.input.reward_table[row].size()) {
                reward_sum += task.input.reward_table[row][option];
                valid_reward += 1.0;
                break;
            }
        }
    }

    std::vector<double> out;
    out.reserve(2 + n_cues);
    out.push_back(1.0);
    out.push_back(valid_reward > 0.0 ? reward_sum / valid_reward : 0.0);

    const double denom = static_cast<double>(
        std::max<std::size_t>(task.input.n_rows, 1)
    );
    for (double count : counts) {
        out.push_back(count / denom);
    }

    return out;
}

std::vector<double> simulated_summary(
    const RunTask& task,
    const RunResult& result
) {
    const std::size_t n_cues = task.behrule.cue.size();
    std::vector<double> counts(n_cues, 0.0);
    double reward_sum = 0.0;
    double acc_sum = 0.0;
    double n_rows = 0.0;

    for (std::size_t row = 0; row < task.input.n_rows; ++row) {
        const std::string& action = result.result.latent[row];
        const auto cue_it = task.behrule.cue_index.find(action);
        if (cue_it != task.behrule.cue_index.end()) {
            counts[cue_it->second] += 1.0;
        }
        if (action == task.input.action[row]) {
            acc_sum += 1.0;
        }
        reward_sum += finite_or(result.result.reward[row], 0.0);
        n_rows += 1.0;
    }

    std::vector<double> out;
    out.reserve(2 + n_cues);
    out.push_back(n_rows > 0.0 ? acc_sum / n_rows : 0.0);
    out.push_back(n_rows > 0.0 ? reward_sum / n_rows : 0.0);

    const double denom = std::max(n_rows, 1.0);
    for (double count : counts) {
        out.push_back(count / denom);
    }

    return out;
}

abcpp::Matrix to_matrix(
    const std::vector<std::vector<double>>& values
) {
    if (values.empty()) {
        return abcpp::Matrix();
    }

    abcpp::Matrix out(values.size(), values.front().size());
    for (std::size_t row = 0; row < values.size(); ++row) {
        for (std::size_t col = 0; col < values[row].size(); ++col) {
            out(row, col) = values[row][col];
        }
    }
    return out;
}

std::vector<abcpp::transform> transformations(
    std::size_t n_params
) {
    return std::vector<abcpp::transform>(
        n_params,
        abcpp::transform::none
    );
}

abcpp::AbcOptions options(
    const ABCControl& control,
    std::size_t n_params
) {
    abcpp::AbcOptions out;
    out.tol = control.tol;
    out.method = abcpp::parse_method(control.method);
    out.kernel = abcpp::parse_kernel(control.kernel);
    out.transformations = transformations(n_params);
    if (!out.transformations.empty()) {
        out.transf = out.transformations.front();
    }
    out.seed = control.seed;
    out.reduction.method = abcpp::parse_reduction(control.reduction);
    out.reduction.n_comp = static_cast<std::size_t>(
        std::max(control.n_comp, 0)
    );
    return out;
}

double sample_param(
    const std::string& name,
    double initial,
    double lower,
    double upper,
    std::mt19937& rng
) {
    if (!std::isfinite(lower) || !std::isfinite(upper)) {
        if (name == "alpha" || name == "lapse" || name == "weight") {
            lower = 0.0;
            upper = 1.0;
        } else if (name == "beta") {
            lower = 0.0;
            upper = 10.0;
        } else {
            const double span = std::max(std::abs(initial), 1.0);
            lower = initial - span;
            upper = initial + span;
        }
    }

    if (upper < lower) {
        std::swap(lower, upper);
    }
    if (upper == lower) {
        return lower;
    }

    std::uniform_real_distribution<double> dist(lower, upper);
    return dist(rng);
}

ABCSummaryStats summary_column(const abcpp::SummaryColumn& col) {
    ABCSummaryStats out;
    out.min = col.min;
    out.q_lower = col.q_lower;
    out.median = col.median;
    out.mean = col.mean;
    out.mode = col.mode;
    out.q_upper = col.q_upper;
    out.max = col.max;
    out.sd = col.sd;
    return out;
}

}  // namespace

ABCSubjectResult estimate_abc(
    const RunTask& raw_task,
    const ABCControl& raw_control
) {
    const ABCControl control = modify_control(raw_control, "abc");

    if (raw_task.params.free_names.empty()) {
        throw std::invalid_argument(
            "estimate_abc requires at least one free parameter."
        );
    }

    ABCSubjectResult out;
    out.subid = task_subid(raw_task);
    out.parameter_names = raw_task.params.free_names;
    out.observed_summary = observed_summary(raw_task);
    out.n_simulations = control.samples;
    out.n_comp_used = control.n_comp;

    std::mt19937 rng(control.seed);
    std::vector<std::vector<double>> param_rows;
    std::vector<std::vector<double>> summary_rows;
    param_rows.reserve(static_cast<std::size_t>(control.samples));
    summary_rows.reserve(static_cast<std::size_t>(control.samples));

    for (int sample = 0; sample < control.samples; ++sample) {
        RunTask sim_task = raw_task;
        sim_task.settings.policy = "on";
        sim_task.params.values["seed"] =
            static_cast<double>(control.seed + sample + 1);

        std::vector<double> param_row;
        param_row.reserve(sim_task.params.free_names.size());

        for (std::size_t index = 0; index < sim_task.params.free_names.size();
             ++index) {
            const std::string& name = sim_task.params.free_names[index];
            const double initial = sim_task.params.get(name);
            const double lower = index < control.lower_bounds.size()
                ? control.lower_bounds[index]
                : -std::numeric_limits<double>::infinity();
            const double upper = index < control.upper_bounds.size()
                ? control.upper_bounds[index]
                : std::numeric_limits<double>::infinity();
            const double value = sample_param(
                name,
                initial,
                lower,
                upper,
                rng
            );
            sim_task.params.values[name] = value;
            param_row.push_back(value);
        }

        const RunResult sim_result = process_MDP_free(sim_task);
        param_rows.push_back(param_row);
        summary_rows.push_back(simulated_summary(sim_task, sim_result));
    }

    abcpp::AbcResult abc_result = abcpp::fit(
        out.observed_summary,
        to_matrix(param_rows),
        to_matrix(summary_rows),
        options(control, raw_task.params.free_names.size())
    );
    abc_result.parameter_names = raw_task.params.free_names;

    const abcpp::SummaryResult summary = abcpp::summary(abc_result);
    out.parameter_summary.reserve(summary.columns.size());
    out.estimates.reserve(summary.columns.size());
    for (const abcpp::SummaryColumn& col : summary.columns) {
        out.parameter_summary.push_back(summary_column(col));
        out.estimates.push_back(col.mean);
    }

    out.accepted_distances = abc_result.distances;
    out.accepted_indices = abc_result.accepted_indices;
    out.n_accepted = static_cast<int>(abc_result.accepted_indices.size());
    out.status = abc_result.status == "ok" ? 1 : -1;
    out.message = abc_result.message;
    return out;
}

std::vector<ABCSubjectResult> estimate_abc(
    const std::vector<RunTask>& tasks,
    const ABCControl& raw_control
) {
    const ABCControl control = modify_control(raw_control, "abc");
    std::vector<ABCSubjectResult> out(tasks.size());

#ifdef _OPENMP
    if (control.threads > 0) {
        omp_set_num_threads(control.threads);
    }
#pragma omp parallel for schedule(dynamic) if(tasks.size() > 1)
#endif
    for (int index = 0; index < static_cast<int>(tasks.size()); ++index) {
        out[static_cast<std::size_t>(index)] =
            estimate_abc(tasks[static_cast<std::size_t>(index)], control);
    }

    return out;
}

}  // namespace multiRL

File: R/R/estimate_abc.R
estimate_abc <- function(
    data,
    id = NULL,
    colnames = list(),
    behrule = list(),
    funcs = list(),
    params = list(),
    priors = list(),
    settings = list(),
    lower = NULL,
    upper = NULL,
    control = list(),
    ...
) {
  .estimate_abc_request(
    data = data,
    id = id,
    colnames = colnames,
    behrule = behrule,
    funcs = funcs,
    params = params,
    priors = priors,
    settings = settings,
    lower = lower,
    upper = upper,
    control = control,
    extra = list(...)
  )
}

.estimate_abc_request <- function(
    data,
    id,
    colnames,
    behrule,
    funcs,
    params,
    priors,
    settings,
    lower,
    upper,
    control,
    extra
) {
  colnames <- .modify_colnames(data = data, colnames = colnames)
  data <- .modify_data_id(
    data = data,
    id = id,
    subid = colnames$subid
  )
  params <- .modify_params(params = params)
  priors <- .modify_priors(
    priors = priors,
    params = params$free
  )
  behrule <- .modify_behrule(behrule = behrule)
  settings <- .modify_settings(settings = settings)
  settings$estimate <- "ABC"
  features <- .modify_features(data = data, colnames = colnames)
  control <- .modify_estimate_abc_control(
    control = control,
    lower = lower,
    upper = upper,
    free_names = base::names(params$free)
  )

  cpp_result <- .estimate_abc(
    object = features$object,
    reward = features$reward,
    action = features$action,
    block = features$block,
    trial = features$trial,
    idinfo = features$idinfo,
    exinfo = features$exinfo,
    cue = behrule$cue,
    rsp = behrule$rsp,
    params = params$flat,
    free_names = base::names(params$free),
    system = settings$system,
    prior_names = priors$name,
    prior_types = priors$type,
    prior_param1 = priors$param1,
    prior_param2 = priors$param2,
    prior_active = priors$active,
    policy = settings$policy,
    name = settings$name,
    mode = settings$mode,
    samples = control$samples,
    tol = control$tol,
    method = control$method,
    reduction = control$reduction,
    n_comp = control$n_comp,
    fake_block = control$fake_block,
    scope = control$scope,
    seed = control$seed,
    threads = control$threads,
    print_level = control$print_level,
    lower_bounds = control$lower_bounds,
    upper_bounds = control$upper_bounds
  )

  out <- list(
    input = list(
      data = data,
      colnames = colnames,
      behrule = behrule,
      funcs = funcs,
      params = params,
      priors = priors,
      settings = settings,
      lower = lower,
      upper = upper,
      control = control,
      features = features,
      extra = extra
    ),
    fit = cpp_result$fit,
    estimator = cpp_result$estimator,
    diagnostics = cpp_result$diagnostics
  )
  base::class(out) <- c("multiRLcpp_estimate_abc", "multiRLcpp_run", "list")
  out
}

.modify_estimate_abc_control <- function(control, lower, upper, free_names) {
  default_control <- list(
    samples = 1000L,
    tol = 0.1,
    method = "rejection",
    reduction = "none",
    n_comp = 0L,
    fake_block = 0L,
    scope = "individual",
    seed = 123L,
    threads = 0L,
    print_level = 1L
  )
  out <- utils::modifyList(default_control, control)
  if (!base::is.null(out$reduce)) {
    out$reduction <- out$reduce
  }
  out$samples <- base::as.integer(out$samples[[1L]])
  out$tol <- base::as.numeric(out$tol[[1L]])
  out$method <- base::as.character(out$method[[1L]])
  out$reduction <- base::as.character(out$reduction[[1L]])
  out$n_comp <- base::as.integer(out$n_comp[[1L]])
  out$fake_block <- base::as.integer(out$fake_block[[1L]])
  out$scope <- base::tolower(base::as.character(out$scope[[1L]]))
  supported_scope <- c("individual", "shared", "universal")
  if (!out$scope %in% supported_scope) {
    base::stop(
      paste0(
        "Unknown ABC scope. Supported scopes are: ",
        base::paste(supported_scope, collapse = ", "),
        "."
      ),
      call. = FALSE
    )
  }
  out$seed <- base::as.integer(out$seed[[1L]])
  out$threads <- base::as.integer(out$threads[[1L]])
  out$print_level <- base::as.integer(out$print_level[[1L]])
  out$lower_bounds <- .modify_estimate_mle_bounds(
    bounds = lower,
    free_names = free_names,
    default = -Inf
  )
  out$upper_bounds <- .modify_estimate_mle_bounds(
    bounds = upper,
    free_names = free_names,
    default = Inf
  )
  out
}
